src/network/networkd-dhcp6.c

```diff
@@ -35,6 +36,88 @@ static bool dhcp6_get_prefix_delegation(Link *link) {
                       RADV_PREFIX_DELEGATION_BOTH);
 }
 
+static bool dhcp6_has_preferred_subnet_id(Link *link) {
+        assert(link->network);
+        return link->network->router_prefix_subnet_id >= 0;
+}
+
+static int dhcp6_get_preferred_delegated_prefix(
+                Manager* manager,
+                Link *link,
+                const struct in6_addr *pd_prefix,
+                uint8_t pd_prefix_len,
+                struct in6_addr *ret_addr) {
+        int r;
+        union in_addr_union pd_prefix_union = { .in6 = *pd_prefix, };
+        const uint8_t prefix_bits = 64 - pd_prefix_len;
```

Comment:
before we do maths like this we should have an assert that validates that pd_prefix_len is indeed <= 64

Reply:
is this actually guaranteed to be <= 64 btw?

Reply:
Yes, we also have a check for that in the calling function:
https://github.com/systemd/systemd/blob/bea37855b18119186e5c03e9e3512b9a08e0511f/src/network/networkd-dhcp6.c#L268

We also have a check that ignores any prefix that is smaller (has greater length) before any of this code is executed: https://github.com/systemd/systemd/blob/bea37855b18119186e5c03e9e3512b9a08e0511f/src/network/networkd-dhcp6.c#L369-L372

Reply:
I'll add another assert there just for good measure in case this changes in the future.